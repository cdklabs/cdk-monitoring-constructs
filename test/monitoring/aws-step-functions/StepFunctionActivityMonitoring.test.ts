import { Duration, Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { Activity } from "aws-cdk-lib/aws-stepfunctions";

import {
  AlarmWithAnnotation,
  StepFunctionActivityMonitoring,
} from "../../../lib";
import { addMonitoringDashboardsToStack } from "../../utils/SnapshotUtil";
import { TestMonitoringScope } from "../TestMonitoringScope";

test("snapshot test: no alarms", () => {
  const stack = new Stack();

  const scope = new TestMonitoringScope(stack, "Scope");

  const activity = new Activity(stack, "Activity", {
    activityName: "DummyActivity",
  });

  const monitoring = new StepFunctionActivityMonitoring(scope, {
    alarmFriendlyName: "DummyActivity",
    activity,
  });

  addMonitoringDashboardsToStack(stack, monitoring);
  expect(Template.fromStack(stack)).toMatchSnapshot();
});

test("snapshot test: all alarms", () => {
  const stack = new Stack();

  const scope = new TestMonitoringScope(stack, "Scope");

  const activity = new Activity(stack, "Activity", {
    activityName: "DummyActivity",
  });

  let numAlarmsCreated = 0;

  const monitoring = new StepFunctionActivityMonitoring(scope, {
    alarmFriendlyName: "DummyActivity",
    activity,
    addDurationP50Alarm: {
      Warning: {
        maxDuration: Duration.minutes(1),
      },
    },
    addDurationP90Alarm: {
      Warning: {
        maxDuration: Duration.minutes(2),
      },
    },
    addDurationP99Alarm: {
      Warning: {
        maxDuration: Duration.minutes(3),
      },
    },
    addFailedActivitiesCountAlarm: {
      Warning: {
        maxErrorCount: 2,
      },
    },
    addFailedActivitiesRateAlarm: {
      Warning: {
        maxErrorRate: 1,
      },
    },
    addTimedOutActivitiesCountAlarm: {
      Warning: {
        maxErrorCount: 2,
      },
    },
    useCreatedAlarms: {
      consume(alarms: AlarmWithAnnotation[]) {
        numAlarmsCreated = alarms.length;
      },
    },
  });

  addMonitoringDashboardsToStack(stack, monitoring);
  expect(numAlarmsCreated).toStrictEqual(6);
  expect(Template.fromStack(stack)).toMatchSnapshot();
});
