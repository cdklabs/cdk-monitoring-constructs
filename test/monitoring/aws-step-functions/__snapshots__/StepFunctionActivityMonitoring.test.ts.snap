// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test: all alarms 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "Activity04690B0A": Object {
      "Properties": Object {
        "Name": "DummyActivity",
      },
      "Type": "AWS::StepFunctions::Activity",
    },
    "ScopeTestDummyActivityDurationP50Warning85BF6D46": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "P50 duration is too long.",
        "AlarmName": "Test-DummyActivity-Duration-P50-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "P50",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ActivityArn",
                    "Value": Object {
                      "Ref": "Activity04690B0A",
                    },
                  },
                ],
                "MetricName": "ActivityRunTime",
                "Namespace": "AWS/States",
              },
              "Period": 300,
              "Stat": "p50",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 60000,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyActivityDurationP90Warning76686B60": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "P90 duration is too long.",
        "AlarmName": "Test-DummyActivity-Duration-P90-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "P90",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ActivityArn",
                    "Value": Object {
                      "Ref": "Activity04690B0A",
                    },
                  },
                ],
                "MetricName": "ActivityRunTime",
                "Namespace": "AWS/States",
              },
              "Period": 300,
              "Stat": "p90",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 120000,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyActivityDurationP99Warning94329F6A": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "P99 duration is too long.",
        "AlarmName": "Test-DummyActivity-Duration-P99-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "P99",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ActivityArn",
                    "Value": Object {
                      "Ref": "Activity04690B0A",
                    },
                  },
                ],
                "MetricName": "ActivityRunTime",
                "Namespace": "AWS/States",
              },
              "Period": 300,
              "Stat": "p99",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 180000,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyActivityFailureCountWarningED994CFB": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Failure count is too high.",
        "AlarmName": "Test-DummyActivity-Failure-Count-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Failed",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ActivityArn",
                    "Value": Object {
                      "Ref": "Activity04690B0A",
                    },
                  },
                ],
                "MetricName": "ActivitiesFailed",
                "Namespace": "AWS/States",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 2,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyActivityFailureRateWarning503C677F": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Failure rate is too high.",
        "AlarmName": "Test-DummyActivity-Failure-Rate-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Failed (avg)",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ActivityArn",
                    "Value": Object {
                      "Ref": "Activity04690B0A",
                    },
                  },
                ],
                "MetricName": "ActivitiesFailed",
                "Namespace": "AWS/States",
              },
              "Period": 300,
              "Stat": "Average",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyActivityTimedOutCountWarningE99C9E4F": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "TimedOut count is too high.",
        "AlarmName": "Test-DummyActivity-TimedOut-Count-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Timeout",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ActivityArn",
                    "Value": Object {
                      "Ref": "Activity04690B0A",
                    },
                  },
                ],
                "MetricName": "ActivitiesTimedOut",
                "Namespace": "AWS/States",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 2,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`snapshot test: no alarms 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "Activity04690B0A": Object {
      "Properties": Object {
        "Name": "DummyActivity",
      },
      "Type": "AWS::StepFunctions::Activity",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
